agent/src/config/mainconfigsource.cpp

```diff
@@ -0,0 +1,247 @@
+        }
+    }
+}
+
+void MainConfigSource::createConfigItems()
+{
+    // TODO: Make this a non member that the above method returns?
+    for (auto &configGroup : m_configsByGroup) {
+        auto group = configGroup.first;
+        auto keys = configGroup.second;
+
+        for (auto &key : keys) {
+            log_debug() << "Before checking type";
+            ConfigType type = typeOf(group, key);
+            log_debug() << "After checking type";
+
+            create(group, key, type);
+        }
+    }
+}
+
+ConfigType MainConfigSource::typeOf(const std::string &group, const std::string &key)
+{
+    return m_typeMapping.at(std::make_pair(group, key));
+}
+
+void MainConfigSource::create(const std::string &group, const std::string &key, ConfigType type)
+{
+    log_debug() << "Creating config " << group << "::" << key;
+
+    switch (type) {
+        case ConfigType::String:
+            log_debug() << "Creating string config";
+            m_stringConfigs.push_back(createStringConfig(group, key));
+            break;
+        case ConfigType::Integer:
+            m_intConfigs.push_back(createIntConfig(group, key));
+            break;
+        case ConfigType::Boolean:
+            m_boolConfigs.push_back(createBoolConfig(group, key));
+            break;
+        default:
+            break;
+    }
+}
+
+StringConfig MainConfigSource::createStringConfig(const std::string &group, const std::string &name)
+{
+    std::string value;
+
+    try {
+        value = _getStringValue(group, name);
+    } catch (ConfigError &error) {
+        log_error() << "Error creating string config";
+        throw;
+    }
+
+    StringConfig config(group, name, value);
+    config.setSource(ConfigSourceType::Main);
+
+    log_debug() << "Constructed StringConfig with key and value: " << name << " " << value;
+
+    return config;
+}
+
+IntConfig MainConfigSource::createIntConfig(const std::string &group, const std::string &name)
+{
+    int value;
+
+    try {
+        value = _getIntegerValue(group, name);
+    } catch (ConfigError &error) {
+        log_error() << "Error creating integer config";
+        throw;
+    }
+
+    IntConfig config(group, name, value);
+    config.setSource(ConfigSourceType::Main);
+
+    log_debug() << "Constructed IntConfig with key and value: " << name << " " << value;
+
+    return config;
+}
+
+BoolConfig MainConfigSource::createBoolConfig(const std::string &group, const std::string &name)
+{
+    bool value;
+
+    try {
+        value = _getBooleanValue(group, name);
+    } catch (ConfigError &error) {
+        log_error() << "Error creating boolean config";
+        throw;
+    }
+
+    BoolConfig config(group, name, value);
+    config.setSource(ConfigSourceType::Main);
+
+    log_debug() << "Constructed BoolConfig with key and value: " << name << " " << value;
+
+    return config;
+}
+
+std::string MainConfigSource::_getStringValue(const std::string &group, const std::string &key) const
```

Comment:
Remains from an old implementation with public getters and internal ones. Will rename it.